### PR TITLE
Raw handling: skip inline shortcodes

### LINKED ISSUE #3806: Pasting inline shortcodes produces separate block
## Issue Overview

When pasting text which includes an inline shortcode, three blocks are produced: a paragraph (text before shortcode), shortcode, and paragraph block (text after shortcode).

## Steps to Reproduce (for bugs)

1. Navigate to Posts > Add New
2. Click into "Write your story" to create a new default block
3. Paste text including an inline shortcode
   - Example: "Due to conflicting travel arrangements of some participants following this past weekend's WordCamp US, we will be canceling the JavaScript Chat for this week. The next meeting will resume as regularly scheduled on [time]Tuesday, December 12, 2017 at 14:00 UTC[/time]."

## Expected Behavior

Since the original text was intended to be contained within a single paragraph, I would expect the inline shortcode to be respected.

## Current Behavior

Excessive blocks are created.

## Screenshots / Video

Inline shortcode paste

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR implements https://github.com/WordPress/gutenberg/issues/3806#issuecomment-381950888.

Fixes #3806.

Only if the shortcode is preceded by a newline (or a plain paragraph tag from the Markdown converter), we should create a shortcode block. Otherwise we should like the shortcode alone as text.

I think this may be a bit too simplistic, but generally goes in the right direction. Maybe we should also check if the shortcode contains any HTML. If it does, maybe it should be "preserved" in a block.

## How has this been tested?
See https://github.com/WordPress/gutenberg/issues/3806#issuecomment-381950888.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Comment by iseulde** [comment]
Also wondering if this should only happen if the transformer is added by the shortcode block. If there is any other block transforming the shortcode, we should probably just let it happens. I think the problem here is the catch-all. Cc @mcsf.

**Comment by iseulde** [comment]
Maybe we come back here to something like `setUnknownShortcodeHandlerName`?

**Comment by mcsf** [comment]
> Also wondering if this should only happen if the transformer is added by the shortcode block. If there is any other block transforming the shortcode, we should probably just let it happens. I think the problem here is the catch-all.

Good question, though I'm not quite convinced. I guess this specific handling of inline shortcodes is mainly in place for backwards compatibility, but also as a way to cope with current limitations of the block paradigm:

1. The goal, foremost, is to prevent breakage of inline flow when converting classic content to blocks. Even when dealing with new content, it's expectable that publishers may need to rely on "legacy" inline shortcodes for certain kinds of data or control (something that perhaps inline tokens will address in the future).

2. Well before Gutenberg, shortcodes have been employed both inline and as "blocks" (on their own line), but I don't see that choice as necessarily bound to the shortcode type. That is, I can see myself using certain shortcodes in both ways:

<table>
<thead>
<tr>
<th>Inline</th>
<th>Block</th>
</tr>
</thead>
<tbody>
<tr>
<td><blockquote>Let's meet at [time]Tuesday, December 12, 2017 at 14:00 UTC[/time].</blockquote></td>
<td><blockquote><h2>Meeting schedule</h2><br><br>[time]Tuesday, December 12, 2017 at 14:00 UTC[/time]</blockquote></td>
</tr></tbody></table>

And, conceptually, when transitioning to Gutenberg, I'd expect the latter scenario to convert to `wp:heading + wp:shortcode`, but the former to remain `wp:paragraph`.

Does this make sense, or was I just rambling? 🙂

**Comment by iseulde** [comment]
@mcsf That makes sense. So what do you think of the handling in this PR?

**Comment by iseulde** [comment]
@aduth @mcsf We discussed this a few times, and I'm not sure what the right solution here is. Shortcakes and Markdown can conflict: `[link](http://...)`. We could do a plain text check first, then make the shortcode parser ignore `[]` followed by `(`? Won't work for reference-style links though. This is the biggest issue.

File: blocks/api/raw-handling/shortcode-converter.js [hunk 1]
```diff
@@ -32,11 +32,22 @@ function segmentHTMLToShortcodeBlock( HTML ) {
...
-	let lastIndex = 0;
 
 	if ( ( match = shortcode.next( transformTag, HTML, lastIndex ) ) ) {
+		const beforeHTML = HTML.substr( 0, match.index );
+
 		lastIndex = match.index + match.content.length;
 
+		// If the shortcode content does not contain HTML and the shortcode is
+		// not on new line (or in paragraph from Markdown converter), consider
+		// the shortcode as inline text.
```
Comment (mcsf) [inline blocks/api/raw-handling/shortcode-converter.js]: I'd add at the end of the sentence: _, and thus skip conversion for this segment._ Does it make sense?

**Review by mcsf — APPROVED** [review]
I'm sure we can fine-tune this in the future if we run into special cases. For now, this is a welcome fix for me.

**Comment by iseulde** [comment]
Thanks for reviewing @mcsf!

**Comment by danielbachhuber** [comment]
@iseulde @mcsf Is there a reason there aren't tests for this pull request (and `segmentHTMLToShortcodeBlock` more generally)?

**Comment by iseulde** [comment]
@danielbachhuber Yes. There's no module yet for the shortcodes, it's still a window global.

**Pull request merged**